rpc/src/cli.rs

```diff
@@ -0,0 +1,64 @@
+use std::io::Write;
+
+use anyhow::Result;
+use clap::{Parser, ValueEnum, ValueHint};
+use rpc::{JerigonRpcClient, NativeRpcClient, RpcClient};
+use url::Url;
+
+#[derive(Parser)]
+pub enum Cli {
+    /// Fetch and generate prover input from the RPC endpoint
+    Fetch {
+        /// The RPC URL
+        #[arg(short = 'u', long, value_hint = ValueHint::Url)]
+        rpc_url: Url,
+        /// The RPC Tracer Type
+        #[arg(short = 't', long, default_value = "jerigon")]
+        rpc_type: RpcType,
+        /// The block number
+        #[arg(short, long)]
+        block_number: u64,
+        /// The checkpoint block number
+        #[arg(short, long, default_value_t = 0)]
+        checkpoint_block_number: u64,
+    },
+}
+
+impl Cli {
+    /// Execute the cli command.
+    pub async fn execute(self) -> Result<()> {
+        match self {
+            Self::Fetch {
+                rpc_url,
+                rpc_type,
+                block_number,
+                checkpoint_block_number,
+            } => {
+                let prover_input = match rpc_type {
+                    RpcType::Jerigon => {
+                        let client = JerigonRpcClient::new(rpc_url);
+                        client
+                            .fetch_prover_input(block_number, checkpoint_block_number)
+                            .await?
+                    }
+                    RpcType::Native => {
+                        let client = NativeRpcClient::new(rpc_url)
+                            .expect("should be able to create native rpc client");
+                        client
+                            .fetch_prover_input(block_number, checkpoint_block_number)
+                            .await?
+                    }
+                };
+                std::io::stdout().write_all(&serde_json::to_vec(&prover_input)?)?;
```

Comment:
> `println!` entails converting the input to a string

This is not true, there are no intermediate allocations.

Also the serde_json::to_writer call comes from yours truly, and again, avoids intermediate allocations

Reply:
> Does the println macro allocate heap memory? I.e. What's the difference (in terms of memory allocation) between calling the format macro to create a string to be printed by println and just formatting things in the println macro?
>
> No, print!, write! and other similar macros that do not return a String do not need to allocate memory; they call format_args! (which is a compiler built-in) to process the arguments and ultimately use std::fmt::write to **write data directly to its destination, without preparing a temporary String.**

Emphasis mine, from https://stackoverflow.com/a/57319140

You can prove this to yourself:
```rust
use core::fmt;

use clap::Parser;

#[derive(clap::Parser)]
struct Repeat {
    count: usize,
}

impl fmt::Display for Repeat {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        for _ in 0..self.count {
            f.write_str("repeat")?
        }
        Ok(())
    }
}

fn main() {
    println!("{}", Repeat::parse());
}
```

```console
$ /usr/bin/time -v ./target/debug/sketchbook 20000 2>&1 | rg 'Maximum resident set size'
#                                            ^^^^^ change this
```

RSS does not vary with input

Reply:
to explain, the above is semantically equivalent to:
```rust
let mut writer = std::io::stdout();
for _ in 0..repeat {
    <std::io::Stdout as std::io::Write>::write_all(&mut writer, b"repeat").unwrap()
}
```

But no-one writes that code, people just use `println!`, `writeln!` etc. 
You can imagine that avoiding intermediate allocations was an explicit goal of these APIs, given that they're used so often, often in environments without an allocator in the case of `core::fmt`

Reply:
> Buffering also has the advantage of batching and iteratively writing, rather than requiring a full allocation of an intermediate type (e.g., a string) before writing out in full.

This is also not true in this case, buffering is completely up to the OS for stdio. Notably, it varies if the output is a TTY or a pipe

Reply:
> I think we probably both agree that to_writer is the more performant solution, and the current std::io::stdout().write_all(&serde_json::to_vec(&prover_input)?)? is a bit odd.

Works for me, let's spend our time elsewhere